net: fix for assert on enabling socket CAN

rbabub opened the pull request:
Socket CAN does not require interface link address to be
assigned, check is not applicable for socket CAN
interfaces. As this address is NULL for socket CAN interface
it results in assertion.

Fixes #37132

Signed-off-by: Ramesh Babu B <ramesh.babu.b@intel.com>

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -3990,7 +3990,11 @@ int net_if_up(struct net_if *iface)
 	 * net_if_set_link_addr(). Better check that now.
 	 */
-	NET_ASSERT(net_if_get_link_addr(iface)->addr != NULL);
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_CAN)) {
```
I don't think that's enough, with such condition we'll skip the LL address verification for any interface, not only the CAN ones.

@alexanderwachter In the corresponding issue, you've mentioned `is_socket_can(iface)`, I don't see such a function in the tree though? Was this just a suggestion to create one or does the function exist somewhere under a different name? I think such a functionality could be achieved with the following code snippet as well?:
```
net_if_l2(iface) == &NET_L2_GET_NAME(CANBUS)
```

alexanderwachter replied:
There is no such function yet. That was just an example as pseudo code

alexanderwachter commented:
Can you add a check if the interface is of type SocketCAN?

rbabub commented:
> Can you add a check if the interface is of type SocketCAN?

ya sure. 

Condition check:
net_if_l2(iface) == &NET_L2_GET_NAME(CANBUS_RAW)

rbabub committed:
net: fix for assert on enabling socket CAN

Socket CAN does not require interface link address to be
assigned, check is not applicable for socket CAN
interfaces. As this address is NULL for socket CAN interface
it results in assertion.

Signed-off-by: Ramesh Babu B <ramesh.babu.b@intel.com>

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -3990,7 +3990,12 @@ int net_if_up(struct net_if *iface)
 	 */
-	NET_ASSERT(net_if_get_link_addr(iface)->addr != NULL);
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_CAN) &&
+	    (net_if_l2(iface) == &NET_L2_GET_NAME(CANBUS_RAW)))	{
```
@alexanderwachter There's `CANBUS_RAW_L2` and `CANBUS_L2` - should the LL address check be skipped for both or is the current solution enough?

alexanderwachter replied:
RAW only. The other is 6loCAN, which is IPv6 over CAN. It has a 14bit address.

jhedberg commented:
Looks alright to me, but why are so many CI checks failing? @rbabub are you looking at those and planning to fix them?

cfriedt commented:
LGTM - please check the compilation errors @rbabub

rbabub commented:
> LGTM - please check the compilation errors @rbabub

ok sure.

rbabub commented:
All checks are passed, could you please review and provide approval.
